src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -156,23 +158,60 @@ TransactionManager::submit_transaction_ertr::future<>
 TransactionManager::submit_transaction(
   TransactionRef t)
 {
-  auto record = cache.try_construct_record(*t);
-  if (!record) {
-    return crimson::ct_error::eagain::make();
-  }
-
   logger().debug("TransactionManager::submit_transaction");
+  return segment_cleaner.do_immediate_work(*t
+  ).safe_then([this, t=std::move(t)]() mutable -> submit_transaction_ertr::future<> {
+    auto record = cache.try_construct_record(*t);
+    if (!record) {
+      return crimson::ct_error::eagain::make();
+    }
 
-  return journal.submit_record(std::move(*record)).safe_then(
-    [this, t=std::move(t)](auto p) mutable {
-      auto [addr, journal_seq] = p;
-      cache.complete_commit(*t, addr, journal_seq);
-      lba_manager.complete_transaction(*t);
-    },
-    submit_transaction_ertr::pass_further{},
-    crimson::ct_error::all_same_way([](auto e) {
-      ceph_assert(0 == "Hit error submitting to journal");
-    }));
+    return journal.submit_record(std::move(*record)).safe_then(
+      [this, t=std::move(t)](auto p) mutable {
+	auto [addr, journal_seq] = p;
+	segment_cleaner.set_journal_head(journal_seq);
+	cache.complete_commit(*t, addr, journal_seq);
+	lba_manager.complete_transaction(*t);
+      },
+      submit_transaction_ertr::pass_further{},
+      crimson::ct_error::all_same_way([](auto e) {
+	ceph_assert(0 == "Hit error submitting to journal");
+      }));
+  });
+}
+
+TransactionManager::get_next_dirty_extents_ret
+TransactionManager::get_next_dirty_extents(journal_seq_t seq)
+{
+  return cache.get_next_dirty_extents(seq);
+}
+
+TransactionManager::rewrite_extent_ret TransactionManager::rewrite_extent(
+  Transaction &t,
+  CachedExtentRef extent)
+{
+  auto FUNC = __func__;
```

Comment:
why?

Reply:
Earlier version had nested lambdas with debugging, this was a way to capture the outermost __func__ and use it elsewhere.  Will fix.